uikit/src/main/java/com/midtrans/sdk/uikit/views/shopeepay/status/ShopeePayStatusActivity.java

```diff
@@ -0,0 +1,343 @@
+            qrCodeContainer = findViewById(R.id.shopeepay_qr_code);
+            qrCodeRefresh = findViewById(R.id.shopeepay_reload_qr_button);
+            setTextColor(qrCodeRefresh);
+            setIconColorFilter(qrCodeRefresh);
+            qrCodeRefresh.setOnClickListener(new OnClickListener() {
+                @Override
+                public void onClick(View v) {
+                    showProgressLayout();
+                    loadQrCode(qrCodeUrl, qrCodeContainer);
+                }
+            });
+            loadQrCode(qrCodeUrl, qrCodeContainer);
+
+            //process expiration
+            expirationText = findViewById(R.id.shopeepay_expiration_text);
+            expirationDesc = findViewById(R.id.shopeepay_expiration_desc);
+
+            String startTime = response.getTransactionTime();
+            if (isExpirationTimeNotAvailable(response) && TextUtils.isEmpty(startTime)) {
+                expirationDesc.setVisibility(View.GONE);
+                expirationText.setVisibility(View.GONE);
+            } else {
+
+                long duration = getPaymentDuration(response);
+                if (duration > 1000) {
+                    setTimer(duration);
+                } else {
+                    expirationText.setVisibility(View.GONE);
+                    expirationDesc.setText(getString(R.string.shopeepay_expiration_expired));
+                }
+            }
+
+            buttonPrimary.setText(getString(R.string.done));
+            buttonPrimary.setOnClickListener(new OnClickListener() {
+                @Override
+                public void onClick(View view) {
+                    showConfirmationDialog(getString(R.string.confirm_shopeepay_qr_scan_tablet));
+                }
+            });
+            buttonPrimary.setTextBold();
+        }
+        textTitle.setText(getString(R.string.shopeepay_status_title));
+    }
+
+    private boolean isExpirationTimeNotAvailable(TransactionResponse response) {
+        return TextUtils.isEmpty(response.getGopayExpiration()) && TextUtils.isEmpty(response.getGopayExpirationRaw());
+    }
+
+    private long getPaymentDuration(TransactionResponse response) {
+        String startTime = response.getTransactionTime();
+        long expiryTimeOnMillis;
+
+        if (TextUtils.isEmpty(response.getGopayExpirationRaw())) {
+            String expirationTime = TextUtils.isEmpty(response.getGopayExpiration()) ? getExpiryTime(response.getTransactionTime()) : response.getGopayExpiration();
+            expiryTimeOnMillis = getDuration(startTime, expirationTime);
+        } else {
+            expiryTimeOnMillis = getDurationByExpirationRaw(startTime, response.getGopayExpirationRaw());
+        }
+
+        return expiryTimeOnMillis;
+    }
+
+    private long getDurationByExpirationRaw(String startTime, String expirationTime) {
+        long startMillis = 0, endMillis = 0;
+        SimpleDateFormat startDf = new SimpleDateFormat("yyyy-MM-dd HH:mm", Locale.US);
```

Comment:
Can you extract this as constant and replace the others too?

Reply:
sure

Reply:
please check again mas